File: src/main/java/net/atos/zac/admin/model/ZaaktypeCmmnMailtemplateParameters.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.admin.model;

import static nl.info.zac.database.flyway.FlywayIntegrator.SCHEMA;

import java.util.Objects;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;

import nl.info.zac.mailtemplates.model.MailTemplate;


@Entity
@Table(schema = SCHEMA, name = "zaaktype_cmmn_mailtemplate_parameters")
@SequenceGenerator(schema = SCHEMA, name = "sq_zaaktype_cmmn_mailtemplate_parameters", sequenceName = "sq_zaaktype_cmmn_mailtemplate_parameters", allocationSize = 1)
public class ZaaktypeCmmnMailtemplateParameters implements UserModifiable<ZaaktypeCmmnMailtemplateParameters> {

    @Id
    @GeneratedValue(generator = "sq_zaaktype_cmmn_mailtemplate_parameters", strategy = GenerationType.SEQUENCE)
    @Column(name = "id")
    private Long id;

    @NotNull @ManyToOne
    @JoinColumn(name = "zaaktype_configuration_id", referencedColumnName = "id")
    private ZaaktypeCmmnConfiguration zaaktypeCmmnConfiguration;

    @NotNull @ManyToOne
    @JoinColumn(name = "id_mail_template", referencedColumnName = "id_mail_template")
    private MailTemplate mailTemplate;

    public Long getId() {
        return id;
    }

    public void setId(final Long id) {
        this.id = id;
    }

    public MailTemplate getMailTemplate() {
        return mailTemplate;
    }

    public void setMailTemplate(final MailTemplate mailTemplate) {
        this.mailTemplate = mailTemplate;
    }

    public ZaaktypeCmmnConfiguration getZaaktypeCmmnConfiguration() {
        return zaaktypeCmmnConfiguration;
    }

    public void setZaaktypeCmmnConfiguration(final ZaaktypeCmmnConfiguration zaaktypeCmmnConfiguration) {
        this.zaaktypeCmmnConfiguration = zaaktypeCmmnConfiguration;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof ZaaktypeCmmnMailtemplateParameters that))
            return false;
        if (mailTemplate == null || that.mailTemplate == null)
            throw new IllegalStateException("mailTemplate is null");
        return Objects.equals(mailTemplate.getId(), that.mailTemplate.getId());
    }

    @Override
    public int hashCode() {
        if (mailTemplate == null)
            throw new IllegalStateException("mailTemplate is null");
        return Objects.hash(mailTemplate.getId());
    }

    @Override
    public boolean isModifiedFrom(ZaaktypeCmmnMailtemplateParameters original) {
        if (mailTemplate == null || original.mailTemplate == null)
            throw new IllegalStateException("mailTemplate is null");
        return Objects.equals(mailTemplate.mail, original.mailTemplate.mail) &&
               !Objects.equals(mailTemplate.getId(), original.mailTemplate.getId());
    }

    @Override
    public void applyChanges(ZaaktypeCmmnMailtemplateParameters changes) {
        mailTemplate = changes.mailTemplate;
    }

    @Override
    public ZaaktypeCmmnMailtemplateParameters resetId() {
        id = null;
        return this;
    }
}

File: src/main/kotlin/nl/info/zac/admin/model/ZaaktypeCmmnMailtemplateParameters.kt
/*
 * SPDX-FileCopyrightText: 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.admin.model

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.SequenceGenerator
import jakarta.persistence.Table
import jakarta.validation.constraints.NotNull
import nl.info.zac.database.flyway.FlywayIntegrator.Companion.SCHEMA
import nl.info.zac.mailtemplates.model.MailTemplate
import nl.info.zac.util.AllOpen
import java.util.Objects

@Entity
@Table(schema = SCHEMA, name = "zaaktype_cmmn_mailtemplate_parameters")
@SequenceGenerator(
    schema = SCHEMA,
    name = "sq_zaaktype_cmmn_mailtemplate_parameters",
    sequenceName = "sq_zaaktype_cmmn_mailtemplate_parameters",
    allocationSize = 1
)
@AllOpen
class ZaaktypeCmmnMailtemplateParameters :
    UserModifiable<ZaaktypeCmmnMailtemplateParameters> {

    @Id
    @GeneratedValue(generator = "sq_zaaktype_cmmn_mailtemplate_parameters", strategy = GenerationType.SEQUENCE)
    @Column(name = "id")
    var id: Long? = null

    @ManyToOne
    @JoinColumn(name = "zaaktype_configuration_id", referencedColumnName = "id")
    @field:NotNull
    var zaaktypeCmmnConfiguration: ZaaktypeCmmnConfiguration? = null

    @ManyToOne
    @JoinColumn(name = "id_mail_template", referencedColumnName = "id_mail_template")
    @field:NotNull
    var mailTemplate: MailTemplate? = null

    @Suppress("ExceptionRaisedInUnexpectedLocation", "UseCheckOrError")
    override fun equals(other: Any?): Boolean {
        if (other !is ZaaktypeCmmnMailtemplateParameters) {
            return false
        }

        val result = mailTemplate?.let { mailtemplate ->
            other.mailTemplate?.let { that ->
                Objects.equals(mailtemplate.id, that.id)
            }
        }
        return result ?: throw IllegalStateException("mailTemplate is null")
    }

    @Suppress("ExceptionRaisedInUnexpectedLocation", "UseCheckOrError")
    override fun hashCode(): Int = mailTemplate?.let { Objects.hash(it.id) } ?: 0

    @Suppress("UseCheckOrError")
    override fun isModifiedFrom(original: ZaaktypeCmmnMailtemplateParameters): Boolean {
        val result = mailTemplate?.let { mailTemplate ->
            original.mailTemplate?.let { otherMailTemplate ->
                Objects.equals(mailTemplate.mail, otherMailTemplate.mail) &&
                    !Objects.equals(mailTemplate.id, otherMailTemplate.id)
            }
        }
        return result ?: throw IllegalStateException("mailTemplate is null")
    }

    override fun applyChanges(changes: ZaaktypeCmmnMailtemplateParameters) {
        mailTemplate = changes.mailTemplate
    }

    override fun resetId(): ZaaktypeCmmnMailtemplateParameters {
        id = null
        return this
    }
}
